Infrastructure/cmake.config

h-2 opened the pull request:
Please have a look @rrahn, ~~but don't merge, yet.~~ and merge!

~~The main open question is: Are we still the package "seqan", but version 3.x (it's like this right now), or are we the package "seqan3"?~~

~~The latter would be a requirement for using both packages in one project I think. But it's less nice...~~

h-2 commented on build/seqan-config.cmake:
```diff
@@ -0,0 +1,364 @@
+#  SEQAN_LIBRARIES
+#  SEQAN_DEFINITIONS
+#
+# The C++ compiler flags that need to be added to SEQAN_CXX_FLAGS:
```
replace with `CMAKE_CXX_FLAGS`

h-2 commented on build/seqan-config.cmake:
```diff
@@ -0,0 +1,364 @@
+# Final check of compilability of seqan3/core/platform.hpp
+# ----------------------------------------------------------------------------
+
+#TODO
```
this will compile an example against the platform.hpp which contains more checks.

h-2 commented:
Lol, travis doesn't do out-of-source builds and conflicts with my newly added `build` subdirectory.

rrahn commented:
!rebuild

rrahn commented:
!rebuild

marehr commented:
@h-2 Didn't I add already this cmake file in a different pr?!

marehr commented:
We think we might have some clash here, because I definitely did part of this work already. I just wonder if I submitted a pull request. 🤔 

I definitely remember that I named the package seqan3 and did some other smaller stuff. Maybe we should coordinate this with my changes that I have in the pipeline.

@h-2 Does this supersedes #117?

h-2 commented:
> @h-2 Didn't I add already add this cmake file in a different pr?!

@marehr None of your open or closed PRs contain this and you weren't there so I went ahead and worked on this...

> @h-2 Does this supersedes #117?

Yes, as stated in #117

marehr commented:
@h-2 Sorry for the trouble, I forgot! I think I wanted to have #112 in before I extracted the common parts of the `benchmark/` and `test/` folder into a `seqan3-config.cmake`.

h-2 commented:
I have just pushed changes and also built in the platform.hpp check which was more complicated than I thought. I can still `s/SEQAN_/SEQAN3_/g` if you prefer.

@marehr Did you miss anything from the file that you had moved their in your branches?

h-2 commented:
now also added c++17 filesystem detection

h-2 commented:
@rrahn I have updated this PR, renamed SEQAN_* to SEQAN3_* and also implemented the discussed functionality to force-require cereal.

Can me merged from my POV.

h-2 commented on build/seqan-config.cmake:
```diff
@@ -0,0 +1,433 @@
+    set (SEQAN_CXX_FLAGS "${SEQAN_CXX_FLAGS} -pthread")
+    # need to explicitly link lpthread with whole-archive because Debian strips bits from the library
+    # that are required for C++14 threading, resulting in broken static builds
+    set (SEQAN_LIBRARIES ${SEQAN_LIBRARIES} "-Wl,--whole-archive -lpthread -Wl,--no-whole-archive")
```
Yes, I will add a TODO

h-2 commented on build/seqan-config.cmake:
```diff
@@ -0,0 +1,433 @@
+# ============================================================================
+#                  SeqAn - The Library for Sequence Analysis
+# ============================================================================
+# Copyright (c) 2006-2018, Knut Reinert, FU Berlin
```
:+1:

h-2 commented on build/seqan-config.cmake:
```diff
@@ -0,0 +1,433 @@
+            OUTPUT_VARIABLE     SEQAN_PLATFORM_TEST_OUTPUT)
+
+if (NOT SEQAN_PLATFORM_TEST)
+    message (STATUS "Performing build and dependency checks in <seqan3/core/platform.hpp> – Failed:\n${SEQAN_PLATFORM_TEST_OUTPUT}")
```
I haven't figured out the exact implications, yet. Strictly speaking everything should be STATUS, I think, because we shouldn't interrupt the build. Instead the build gets interrupted when we return without setting SEQAN3_FOUND. **iff** it was declared as required. Otherwise there shouldn't be en error, because the module is optional anyway.

h-2 commented on test/CMakeLists.txt:
```diff
@@ -1,24 +1,104 @@
+set (CMAKE_INCLUDE_PATH ${SEQAN3_ROOT}/include ${CMAKE_INCLUDE_PATH})
+
+# SeqAn cmake module
+set (CMAKE_PREFIX_PATH ${SEQAN3_ROOT}/build ${CMAKE_PREFIX_PATH})
```
I wanted to have one directory where we have all the config files for build system, i.e. cmake, pkgconfig and possibly mason. How would you call it? `util` is too generic and `cmake_module` is too specific.

h-2 commented:
I have done substantial changes to address the concerns. workaround cmake akwardness, make checks pass where SeqAn is not required and possibly the "search" is quite; and in general make the detection process display nice diagnostics instead of the defaults.

I expect failing check because of broken upstream gtest. These can be addressed with the other PR.

rrahn commented on build_system/seqan3-config.cmake:
```diff
@@ -0,0 +1,527 @@
+    #error NOCXX17
+    #endif
+    int main() {}")
+check_cxx_source_compiles("${CXXSTD_TEST_SOURCE}" CXX17_BUILTIN)
```
I think spaces between these statements make it a little bit more readable.
Like before an if-clause begins, insert a new line.
If a multi line statement ends insert a new line.

rrahn commented on test/CMakeLists.txt:
```diff
@@ -1,24 +1,101 @@
+# ============================================================================
+#                  SeqAn - The Library for Sequence Analysis
+# ============================================================================
+# Copyright (c) 2006-2018, Knut Reinert, FU Berlin
```
Please update the header.
